pageserver/src/tenant/remote_timeline_client/upload.rs

```diff
@@ -120,6 +120,30 @@ pub(super) async fn upload_timeline_layer<'a>(
         .with_context(|| format!("upload layer from local path '{source_path}'"))
 }
 
+pub(super) async fn copy_timeline_layer(
+    conf: &'static PageServerConf,
+    storage: &GenericRemoteStorage,
+    source_path: &Utf8Path,
+    source_metadata: &LayerFileMetadata,
+    target_path: &Utf8Path,
+    target_metadata: &LayerFileMetadata,
+    cancel: &CancellationToken,
+) -> anyhow::Result<()> {
+    fail_point!("before-copy-layer", |_| {
+        bail!("failpoint before-copy-layer")
+    });
+
+    pausable_failpoint!("before-copy-layer-pausable");
+
+    let source_path = remote_path(conf, source_path, source_metadata.generation)?;
+    let target_path = remote_path(conf, target_path, target_metadata.generation)?;
+
+    storage
+        .copy_object(&source_path, &target_path, cancel)
+        .await
+        .with_context(|| format!("adopt layer {source_path} to {target_path}"))
```

Comment:
`Adopt` is higher-level context than this method.

```suggestion
        .with_context(|| format!("copy layer {source_path} to {target_path}"))
```

Reply:
28515eb5a